tqftpserv.c

```diff
@@ -98,8 +99,21 @@ static ssize_t tftp_send_data(struct tftp_client *client,
 
 	p += len;
 
-	// printf("[TQFTP] Sending %zd bytes of DATA\n", p - buf);
-	len = send(client->sock, buf, p - buf, 0);
+	/* If rsize was set, we should limit the data in the response to n bytes */
+	if (rsize != 0) {
+		/* Header (4 bytes) + data size */
+		send_len = 4 + rsize;
+		if (send_len > p - buf) {
+			printf("[TQFTP] requested data of %ld bytes but only read %ld bytes from file, rejecting\n", rsize, len);
+			free(buf);
+			return -EINVAL;
+		}
+	} else {
+		send_len = p - buf;
+	}
+
```

Comment:
Can we change this to use `rsize` to limit `clk->blksize` instead? Also, is this an error to read less than rsize?

Last but not least, I see that this rsize is not the same as the rsize set by the client. I must admit, it caused confusion on my side. Could you please rename the param here?

Reply:
> Can we change this to use rsize to limit clk->blksize instead?

I assume you mean reading not all of `client->blksize` in `pread` if less bytes will actually be sent? The current code actually matches what the proprietary tftp_server is doing, it's always reading the full block and only sending x bytes back. Not sure if there's any good reason for that or just nobody there noticed.

> Also, is this an error to read less than rsize?

Since there's of course no public docs for this, I can't say for sure what the spec would say.

But I haven't seen a use case where this would be required. So that's why I've been conservative here and return an error in that case. If we find a use case where that's required we can just patch tqftpserv again.

Reply:
Ack for the error. 
Yes, reading less data in pread can simplify the code of the wrapping function.

Reply:
Reading less is somewhat connected to here https://github.com/linux-msm/tqftpserv/pull/13#discussion_r1544602606 , I don't see a straight foward way to have that..

Also honestly I'm a bit afraid of changing the logic so much again, not a fan of user-space C ;)

Reply:
Sorta, if rsize option is not set we'll just send everything until the end of the file.
For rsize being set it's really my interpretation of it but it has held up so far. And I'd also think it's logical that if the client wants x bytes we should send x bytes?